pkg/instrumentation/annotationmutator.go

```diff
@@ -0,0 +1,106 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package instrumentation
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// AnnotationMutation is used to modify an annotation map.
+type AnnotationMutation interface {
+	// Mutate attempts to modify the annotations map. Returns whether the function changed the annotations.
+	Mutate(annotations map[string]string) bool
+}
+
+type insertAnnotationMutation struct {
+	insert     map[string]string
+	requireAll bool
+}
+
+func (m *insertAnnotationMutation) Mutate(annotations map[string]string) bool {
+	if m.requireAll && !m.validate(annotations) {
+		return false
+	}
+	var mutated bool
+	for key, value := range m.insert {
+		if _, ok := annotations[key]; !ok {
```

Comment:
do we need a mutex for read/write of the annotations map?

Reply:
Only if we're modifying the annotations on multiple threads. We aren't currently doing that, so I don't think we need it yet.